src/EditorFeatures/Core/Classification/CopyPasteAndPrintingClassificationBufferTaggerProvider.Tagger.cs

```diff
@@ -150,18 +159,50 @@ public IEnumerable<ITagSpan<IClassificationTag>> GetAllTags(NormalizedSnapshotSp
-                (cachedTags, spans),
-                _: (ITagSpan<IClassificationTag>?)null);
-        }
+                (cachedTags, spans));
 
-        private Task ProduceTagsAsync(
-            TaggerContext<IClassificationTag> context, DocumentSnapshotSpan snapshotSpan,
-            IClassificationService classificationService, ClassificationOptions options, ClassificationType type, CancellationToken cancellationToken)
-        {
-            return ClassificationUtilities.ProduceTagsAsync(
-                context, snapshotSpan, classificationService, _owner._typeMap, options, type, cancellationToken);
+            async ValueTask AddSyntacticSpansAsync(NormalizedSnapshotSpanCollection spans, SegmentedList<ITagSpan<IClassificationTag>> result, bool _)
+            {
+                Contract.ThrowIfTrue(spans.Count != 1, "We should only be asking for a single span when getting the syntactic classifications");
+
+                await AddSpansAsync(spans, result,
+                    span => classificationService.AddSyntacticClassificationsAsync(document, span, tempClassifiedSpans, cancellationToken)).ConfigureAwait(false);
+            }
+
+            async ValueTask AddSemanticSpansAsync(NormalizedSnapshotSpanCollection spans, SegmentedList<ITagSpan<IClassificationTag>> result, bool _)
+            {
+                Contract.ThrowIfTrue(spans.Count != 1, "We should only be asking for a single span when getting the semantic classifications");
+
+                await AddSpansAsync(spans, result,
+                    span => classificationService.AddSemanticClassificationsAsync(document, span, options, tempClassifiedSpans, cancellationToken)).ConfigureAwait(false);
+            }
+
+            async ValueTask AddEmbeddedSpansAsync(NormalizedSnapshotSpanCollection stringLiteralSpans, SegmentedList<ITagSpan<IClassificationTag>> result, bool _)
+            {
+                // Note: many string literal spans may be passed in here.
+                await AddSpansAsync(stringLiteralSpans, result,
+                    span => classificationService.AddEmbeddedLanguageClassificationsAsync(document, span, options, tempClassifiedSpans, cancellationToken)).ConfigureAwait(false);
+            }
+
+            async ValueTask AddSpansAsync(
+                NormalizedSnapshotSpanCollection spans,
+                SegmentedList<ITagSpan<IClassificationTag>> result,
+                Func<TextSpan, Task> addAsync)
+            {
+                Contract.ThrowIfTrue(tempClassifiedSpans.Count != 0);
+
+                foreach (var span in spans)
+                {
+                    await addAsync(span.Span.ToTextSpan()).ConfigureAwait(false);
+
+                    foreach (var classifiedSpan in tempClassifiedSpans)
+                        result.Add(ClassificationUtilities.Convert(_owner._typeMap, snapshot, classifiedSpan));
+
+                    tempClassifiedSpans.Clear();
```

Comment:
>tempClassifiedSpans.Clear(); [](http://example.com/codeflow?start=20&length=28)

I would find this easier to read if the Classifier.GetPooledList was done inside this loop and passed as a param to the addAsync call.